File: 2023/18/18.rb
#!/usr/bin/env ruby

part1_path = []
part2_path = []

NEIGHBORS = {
  'R' => [1, 0],
  'D' => [0, 1],
  'L' => [-1, 0],
  'U' => [0, -1],
}

$stdin.readlines.lazy.map(&:chomp).each do |line|
  direction, distance, color = line.split(' ')
  part1_path << [direction, distance.to_i]
  hex = color.tr('#()', '')
  distance, direction = hex[0..-2].to_i(16), NEIGHBORS.keys[hex[-1].to_i]
  part2_path << [direction, distance]
end

def print_grid(grid)
  grid.each do |row|
    puts row.join('')
  end
  puts
end

def find_vertices(path)
  x, y = [0, 0]
  vertices = [[x, y]]
  path.each do |dir, dist|
    x, y = [x, y].zip(NEIGHBORS[dir].map { |d| d * dist }).map(&:sum)
    vertices << [x, y]
  end
  vertices
end

def area(vertices, path)
  sum = vertices.each_cons(2).sum do |(v1_x, v1_y), (v2_x, v2_y)|
    v1_x * v2_y - v1_y * v2_x
  end + (vertices.last.first * vertices.first.last - vertices.last.last * vertices.first.first)
  (sum.abs + path.sum { _2 }) / 2 + 1 # The perimeter needs to be added to the sum, but I am not actually sure where the +1 comes from
end

puts area(find_vertices(part1_path), part1_path)
puts area(find_vertices(part2_path), part2_path)

File: 2023/18/18.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Step struct {
	direction byte
	distance  int
}

type V struct {
	x int
	y int
}

func ParseInput() ([]Step, []Step) {
	scanner := bufio.NewScanner(os.Stdin)
	part1Path, part2Path := []Step{}, []Step{}
	directions := []byte{'R', 'D', 'L', 'U'}
	for scanner.Scan() {
		path_parts := strings.Split(scanner.Text(), " ")
		directionStr, distanceStr, color := path_parts[0], path_parts[1], path_parts[2]
		part1Dist, _ := strconv.Atoi(distanceStr)
		part1Path = append(part1Path, Step{directionStr[0], part1Dist})
		part2Dist, _ := strconv.ParseInt(color[2:7], 16, 64)
		part2DirInt, _ := strconv.Atoi(string(color[7]))
		part2Dir := directions[part2DirInt]
		part2Path = append(part2Path, Step{part2Dir, int(part2Dist)})
	}
	return part1Path, part2Path
}

func runStep(x int, y int, step Step) (int, int) {
	switch step.direction {
	case 'R':
		return x + step.distance, y
	case 'D':
		return x, y + step.distance
	case 'L':
		return x - step.distance, y
	case 'U':
		return x, y - step.distance
	}
	return 0, 0
}

func findVertices(path []Step) []V {
	x, y := 0, 0
	vertices := []V{{x, y}}

	for _, step := range path {
		x, y = runStep(x, y, step)
		vertices = append(vertices, V{x, y})
	}

	return vertices
}

func area(vertices []V, path []Step) int {
	sum := 0
	lastVIndex := len(vertices) - 1
	for i := range lastVIndex {
		v1, v2 := vertices[i], vertices[i+1]
		sum += (v1.x * v2.y) - (v1.y * v2.x)
	}
	sum += (vertices[lastVIndex].x * vertices[0].y) - (vertices[lastVIndex].y * vertices[0].x)
	sum = int(math.Abs(float64(sum)))
	for _, s := range path {
		sum += s.distance
	}
	return sum/2 + 1
}

func main() {
	part1Path, part2Path := ParseInput()

	fmt.Println(area(findVertices(part1Path), part1Path))
	fmt.Println(area(findVertices(part2Path), part2Path))
}
